### PR TITLE
SocketAPI: Make it easier to add or remove item in the action menu

### PR DESCRIPTION
**Pull request by ogoffart**
By making it dynamic.
So far only the nautilus and dolphin shell extension have been ported

### CODE REVIEW

**Comment by michaelstingl** [comment]
sounds 🆒  😎

**Comment by ogoffart** [comment]
The reason to do that is that the new placeholder change might want to add entries in this menu, and so it would be much easier if we can change easily the entries without having to modify every shell extension.

So far I only did the change to nautilus and dolphin. Waiting for feedback before doing the rest.

The filter parametter is a regexp to on the whole filename.   And the filename is made to ends with '/' for directories.
So we can have special entries for directories, or .owncloud files.
I realized when porting nautilus that it only add entries on files that have the OK status, this might be a problem and something we might want to encode in the filter instead.

Also i just tought perhaps there should be a way to set the icon

**Comment by SamuAlfageme** [comment]
Some relevant tickets that'd benefit from this: https://github.com/owncloud/client/issues/5903 (if I understood correctly) and, on a deeper level: https://github.com/owncloud/client/issues/6292

**Comment by ogoffart** [comment]
@SamuAlfageme  Yes, this would help for all these issue.

There are some issues about graying out the menu for files that can be shared and so on. So I need to brainstorm with @ckamm on smarter filter.
An idea i had was that we could add one more string in the "STATUS:" replies or update and filter on that instead of on the name. That would work fine for Nautilus in which all extension are the same. That would somehow work for windows because we do synchronous query anyway we can as well query the status.
But for Dolphin that might be harder because the action plugin don't have access to the status right now.

We also need to encode in the filter if we should disable or remove the options.
Also some options might be valid if there is a multiple selection.

**Comment by ckamm** [comment]
@SamuAlfageme @ogoffart To make it easier to think about this, here's a list of behaviors we want:
* Placeholder files should have a "Download this file" menu action, not others.
* Files (not directories) might have a "Remove local copy, keep remote" menu action to convert back to placeholder only.
* If the server has sharing disabled the "Share..." action should be hidden.
* For files that can't be shared the "Share..." action should be disabled. (Problem: User doesn't understand why)
* If the client isn't connected the "Share..." action should be disabled. (Problem: User doesn't understand why)
* It should be easy to add new simple actions that just trigger some behavior in the client.
* Maybe multiaccount issues - but it's so far out of usual user behavior that I'd not worry overmuch

I did not fully get whether #4608 and #6252 are related too.

File: shell_integration/windows/OCContextMenu/OCClientInterface.cpp [hunk 1]
```diff
@@ -45,7 +45,8 @@ OCClientInterface::ContextMenuInfo OCClientInterface::FetchInfo()
     if (!socket.Connect(pipename)) {
         return {};
     }
-    socket.SendMsg(L"GET_STRINGS\n");
+    socket.SendMsg(L"GET_STRINGS:CONTEXT_MENU_TITLE\n");
```
Comment (ckamm) [inline shell_integration/windows/OCContextMenu/OCClientInterface.cpp]: I don't think you can request a particular string only right now.

File: shell_integration/windows/OCUtil/StringUtil.h [hunk 2]
```diff
@@ -61,6 +61,30 @@ class __declspec(dllexport) StringUtil {
...
+        auto statusBegin = source.find(L':', 0);
+        assert(statusBegin != std::wstring::npos);
+
+        auto statusEnd = source.find(L':', statusBegin + 1);
+        if (statusEnd == std::wstring::npos) {
+            // the command do not contains two colon?
+            return false;
+        }
+
+        auto thirdCollon = source.find(L':', statusEnd + 1);
```
Comment (ckamm) [inline shell_integration/windows/OCUtil/StringUtil.h]: thirdColon

Comment (ogoffart) [reply shell_integration/windows/OCUtil/StringUtil.h]: will fix

File: src/gui/socketapi.cpp [hunk 3]
```diff
@@ -545,11 +545,37 @@ void SocketApi::command_GET_STRINGS(const QString &, SocketListener *listener)
     } };
     listener->sendMessage(QString("GET_STRINGS:BEGIN"));
     for (auto key_value : strings) {
-        listener->sendMessage(QString("STRING:%1:%2").arg(key_value.first, key_value.second));
+        if (argument.isEmpty() || argument == QLatin1String(key_value.first)) {
```
Comment (ogoffart) [inline src/gui/socketapi.cpp]: @ckamm:  This change allow to request a special string.

File: shell_integration/windows/OCContextMenu/OCClientInterface.cpp (hunk 1 above)
Comment (ogoffart) [reply shell_integration/windows/OCContextMenu/OCClientInterface.cpp]: See the change in SocketApi::command_GET_STRINGS

File: src/gui/socketapi.cpp [hunk 4]
```diff
@@ -545,11 +545,37 @@ void SocketApi::command_GET_STRINGS(const QString &, SocketListener *listener)
...
+            listener->sendMessage(QString("STRING:%1:%2").arg(key_value.first, key_value.second));
+        }
     }
     listener->sendMessage(QString("GET_STRINGS:END"));
 }
 
+void SocketApi::command_GET_MENU_ITEMS(const QString &argument, OCC::SocketListener *listener)
+{
+    listener->sendMessage(QString("GET_MENU_ITEMS:BEGIN"));
+    bool hasSeveralFiles = argument.contains(QLatin1Char('\x1e')); // Record Separator
```
Comment (guruz) [inline src/gui/socketapi.cpp]: So while the extensions do support selecting and then querying multiple files, the client does not yet do I understand this correctly? @ogoffart @ckamm

Comment (ogoffart) [reply src/gui/socketapi.cpp]: The placeholder branch does.

Comment (ckamm) [reply src/gui/socketapi.cpp]: Yep, you can ask to download multiple placeholders at once with this functionality :)

**Pull request merged**